File: python/hand_identifier.py
class HandIdentifier:
    def assign_type(hand):
        pass

    def is_royal_straight_flush(cards):
        if HandIdentifier.is_royal_straight(cards) and HandIdentifier.is_flush(cards):
            return True

        return False

    def is_royal_straight(cards):
        if cards[0].get_value() == 0 and cards[1].get_value() == 9 and cards[2].get_value() == 10 and cards[3].get_value() == 11 and cards[4].get_value() == 12:
            return True

        return False

    def is_straight_flush(cards):
        if HandIdentifier.is_straight(cards) and HandIdentifier.is_flush(cards):
            return True

        return False

    def is_four_of_a_kind(cards):
        if 4 in HandIdentifier.get_frequency_set(cards):
            return True

        return False

    def is_full_house(cards):
        frequency_set = HandIdentifier.get_frequency_set(cards)

        if 3 in frequency_set and 2 in frequency_set:
            return True

        return False

    def is_flush(cards):
        prev_suit = cards[0].get_suit()

        for card in cards:
            if card.get_suit() != prev_suit:
                return False

            prev_suit = card.get_suit()

        return True

    def is_straight(cards):
        if HandIdentifier.is_royal_straight(cards):
            return True

        prev_value = cards[0].get_value() - 1

        for card in cards:
            curr_value = card.get_value()

            if curr_value != prev_value + 1:
                return False

            prev_value = curr_value

        return True

    def is_three_of_a_kind(cards):
        if 3 in HandIdentifier.get_frequency_set(cards):
            return True

        return False

#=============== Helper methods ===============  
    def get_frequency_set(cards):
        frequency_set = []

        prev_value = -1
        for card in cards:
            if card.get_value() == prev_value:
                continue

            frequency_set.append(cards.count(card))
            prev_value = card.get_value()

        return sorted(frequency_set)

File: go/hand_identifier.go
package main

import (
    "sort"
)

func AssignType(h *Hand) {
    cards := h.GetSortedCards()

    if IsRoyalStraightFlush(cards) {
        h.handType = 9
    } else if IsStraightFlush(cards) {
        h.handType = 8
    } else if IsFourOfAKind(cards) {
        h.handType = 7
    } else if IsFullHouse(cards) {
        h.handType = 6
    } else if IsFlush(cards) {
        h.handType = 5
    } else if IsStraight(cards) {
        h.handType = 4
    } else if IsThreeOfAKind(cards) {
        h.handType = 3
    } else {
        h.handType = 0
    }
}

func IsRoyalStraightFlush(cards []Card) bool {
    if IsRoyalStraight(cards) && IsFlush(cards) {
        return true
    }

    return false
}

func IsRoyalStraight(cards []Card) bool {
    if cards[0].value == 0 && cards[1].value == 9 && cards[2].value == 10 && cards[3].value == 11 && cards[4].value == 12 {
        return true
    }

    return false
}

func IsStraightFlush(cards []Card) bool {
    if IsStraight(cards) && IsFlush(cards) {
        return true
    }

    return false
}


func IsFourOfAKind(cards []Card) bool {
    set := GetFrequencySet(cards)

    for _, v := range set {
        if v == 4 {
            return true
        }
    }

    return false
}

func IsFullHouse(cards []Card) bool {
    set := GetFrequencySet(cards)

    if set[0] == 2 && set[1] == 3 {
        return true
    }

    return false
}

func IsFlush(cards []Card) bool {
    prevSuit := cards[0].suit

    for _, card := range cards {
        if card.suit != prevSuit {
            return false
        }
    }

    return true
}

func IsStraight(cards []Card) bool {
    if IsRoyalStraight(cards) {
        return true
    }

    prevValue := cards[0].value - 1

    for _, card := range cards {
        currValue := card.value

        if currValue != prevValue + 1 {
            return false
        }

        prevValue = currValue
    }

    return true
}

func IsThreeOfAKind(cards []Card) bool {
    set := GetFrequencySet(cards)

    for _, v := range set {
        if v == 3 {
            return true
        }
    }

    return false
}

// =============== Helper ===============

func GetFrequencySet(cards []Card) []int {
    set := []int{}

    prevValue := -1

    for _, card := range cards {
        currValue := card.value

        if currValue == prevValue {
            continue
        }

        count := 0

        for _, c := range cards {
            if c.value == currValue {
                count += 1
            }
        }

        set = append(set, count)
        prevValue = currValue
    }

    sort.Ints(set)

    return set
}
